index.js

```diff
@@ -93,121 +114,153 @@ rl.on('line', function(line) {
+            console.log('Creating admin user', admin.forename, '......');
+            createUser(admin, masterToken).then(result => {
+                console.log('Admin user created with username:', result.username, 'id:', result.id);
 
-                        function handleAddUserToGroup(user) {
-                            console.log('Adding user (' + user.forename + ') to group ' + groupData.name + '......');
-                            return request('PUT', '/groups/' + groupId + '/members/' + userIdMap.get(user.phoneNumber),
-                                {'Authorization': 'Bearer ' + token});
+                console.log('Logging in admin user......');
+                login(admin).then(result => {
+                    console.log('token for admin user:', result.token);
+                    token = result.token;
+
+                    console.log('Creating group ' + groupName, 'with token:', masterToken, '.......');
+                    createGroup(groupName, token).then(result => {
+                        var groupId = result.id;
+                        console.log('group created, id:', groupId);
+
+                        function handleCreateUser(user) {
+                            console.log('Creating user ' + user.forename + '......');
+                            return createUser(user, masterToken);
                         }
-                        console.log('Adding users to group......');
-                        var addUserToGroupPromises = users.map(handleAddUserToGroup);
-                        Promise.all(addUserToGroupPromises).then(results => {
-                            console.log('Users successfully added to group');
-                            persistToken(token);
-                            console.log('Web token successfully persisted to file');
-                            resolve();
+
+                        console.log('Creating users......');
+                        var createUserPromises = users.map(handleCreateUser);
+                        var userIdMap = new Map();
+                        Promise.all(createUserPromises).then(results => {
+                            results.forEach(result => {
+                                userIdMap.set(result.username, result.id);
+                                console.log('User created with username:', result.username, 'id:', result.id);
+                            });
+
+                            function handleAddUserToGroup(user) {
+                                console.log('Adding user (' + user.forename + ') to group ' + groupName + '......');
+                                return addToGroup(groupId, userIdMap.get(user.phoneNumber), token);
+                            }
+
+                            console.log('Adding users to group......');
+                            var addUserToGroupPromises = users.map(handleAddUserToGroup);
+                            Promise.all(addUserToGroupPromises).then(results => {
```

Comment:
I think here you could `return Promise.all(addUserToGroupPromises)` and somewhere up the chain return the token, so that the `persistToken` action could be done at the first level of promises, in a place where we can more easily see it.